src/app/types/filter-definition.ts

```diff
@@ -0,0 +1,88 @@
+import { FilterValueOptions, FiltersAnd } from '@app/types/filters';
+/**
+ *
+ * `for` and `field` are used to identify the filter.
+ */
+export type FilterDefinitionRootString<T extends number> = {
+  for: 'root';
+  field: T;
+  type: 'string'
+};
+  
+/**
+   *
+   * `for` and `field` are used to identify the filter.
+   */
+export type FilterDefinitionRootNumber<T extends number> = {
+  for: 'root';
+  field: T;
+  type: 'number'
+};
+  
+/**
+   *
+   * `for` and `field` are used to identify the filter.
+   */
+export type FilterDefinitionRootArray<T extends number> = {
+  /**
+     * Used to know which field comes from since it's just a number from an enum.
+     */
+  for: 'root';
+  field: T;
+  type: 'array';
+};
+  
+/**
+   *
+   * `for` and `field` are used to identify the filter.
+   */
+export type FilterDefinitionRootStatus<T extends number> = {
+  /**
+     * Used to know which field comes from since it's just a number from an enum.
+     */
+  for: 'root';
+  field: T;
+  type: 'status';
+  statuses: FilterValueOptions;
+};
+  
+export type FilterDefinitionRootDate<T extends number> = {
+  /**
+   * Used to know which field comes from since it's just a number from an enum.
+   */
+  for: 'root';
+  field: T;
+  type: 'date';
+};
+
+  
+
+
+export type FilterDefinitionTaskOptionString<T extends number | null> = {
+  /**
+   * Used to know which field comes from since it's just a number from an enum.
+   */
+  for: 'options';
+  field: T;
+  type: 'string';
+};
+
+
+type FilterDefinitionRoot<T extends number> = FilterDefinitionRootString<T> | FilterDefinitionRootNumber<T> | FilterDefinitionRootArray<T> | FilterDefinitionRootStatus<T> | FilterDefinitionRootDate<T>;
+
+
+export type FilterDefinitionTaskOption<T extends number | null> = FilterDefinitionTaskOptionString<T>;
+
+export type FilterDefinition<T extends number, U extends number | null = null> = FilterDefinitionRoot<T> | FilterDefinitionTaskOption<U>;
+
+export type FilterFor<T extends number, U extends number | null = null> = FilterDefinition<T, U>['for'];
+
+
+
+export abstract class BasicFilterService {
```

Comment:
DataFilterService might be a better name for this class.

After all, it handles everything data related to the filters, reseting them, restoring them...

Reply:
I agree with you I'll rename it